src/layers/domain/core/cpm_system_id/v1.py

```diff
@@ -1,86 +1,139 @@
+        return CpmSystemIdType.ASIDNUMBER
+
+    @staticmethod
+    def _format_key(latest_id: int) -> str:
+        """Return the ASID as a 12-digit number."""
+        return f"{latest_id:012d}"
 
     @classmethod
     def validate_key(cls, key: str) -> bool:
-        """
-        Validate that the ASID has the correct format.
-        """
+        """Validate that the ASID has the correct format."""
         return key.isdigit() and len(key) == 12 and key.startswith("2")
 
 
 class PartyKeyId(CpmSystemId):
     ods_code: str
+    latest: Optional[int]
 
     @classmethod
-    def create(cls, current_id=None, ods_code: str = ""):
-        # Initialize current_id with a default if not provided
-        current_id = current_id["latest"] if current_id is not None else 849999
-        instance = cls(
-            key_name="PARTYNUMBER",
-            current_number=current_id,
-            latest=current_id + 1,
-            ods_code=ods_code,
-            latest_id="",  # Temporary placeholder for latest_id
+    def create(cls, current_id=None, ods_code=""):
+        current_id = current_id.get("latest") if current_id else 849999
+        instance = cls(key_name=cls._get_key_name(), latest_id="", ods_code=ods_code)
+        instance.latest = current_id + 1
+        instance.latest_id = cls._format_key(
+            ods_code=ods_code, latest_id=current_id + 1
         )
-        instance.latest_id = instance._format_key()
         return instance
 
-    def _format_key(self) -> str:
-        """
-        Format the party key with the ODS code and a 7-digit number.
-        """
-        return f"{self.ods_code}-{self.latest:06d}"
+    @classmethod
+    def _get_key_name(cls) -> CpmSystemIdType:
+        return CpmSystemIdType.PARTYKEYNUMBER
+
+    @staticmethod
+    def _format_key(ods_code: str, latest_id: int) -> str:
+        """Format the party key with the ODS code and a 7-digit number."""
+        return f"{ods_code}-{latest_id:06d}"
 
     @classmethod
     def validate_key(cls, key: str) -> bool:
-        """
-        Validate that the party key has the correct format.
-        """
+        """Validate that the party key has the correct format."""
         parts = key.split("-")
         if len(parts) != 2:
             return False
         ods_code, number = parts
         return ods_code.isalpha() and number.isdigit() and len(number) == 6
+
+
+class ProductId(CpmSystemId):
+
+    @classmethod
+    def create(cls):
+        """No current_id needed, key is generated randomly."""
+        instance = super().create()
+        instance.latest_id = cls._format_key()
+        return instance
+
+    @classmethod
+    def _get_key_name(cls) -> CpmSystemIdType:
+        return CpmSystemIdType.PRODUCTID
+
+    @classmethod
+    def _format_key(cls) -> str:
+        """Generate the product ID as a random string in the format 'P.XXX-XXX'."""
+        PART_LENGTH = 3
+        N_PARTS = 2
+        PRODUCT_ID_VALID_CHARS = "ACDEFGHJKLMNPRTUVWXY34679"
```

Comment:
You have this in two separate methods on the class - might be worth pulling it out as a productID class variable? Especially since its a specific collection of characters its something that might be faffy to track if its in more than one place